M2/Macaulay2/packages/Style.m2

```diff
@@ -3,16 +3,182 @@ newPackage( "Style",
      AuxiliaryFiles => true,
      Headline => "style sheets and images for the documentation",
      Keywords => {"Documentation"},
-     Version => "1.0"
+     Version => "1.1"
      )
 
+---------------------------------------------
+-- grammar generation code adapted from    --
+-- M2/Macaulay2/editors/make-M2-symbols.m2 --
+---------------------------------------------
+
+export {"generateGrammar"}
+
+importFrom(Core, "sortBy")
+
+is := X -> (name, symb) -> instance(value symb, X)
+
+isAlpha        := s -> match("^[[:alpha:]]+$", s)
+isAlphaNumeric := s -> match("^[[:alnum:]]+$", s)
+
+-- Things we want to highlight:
+isType     := is Type
+isKeyword  := is Keyword
+isFunction := is Function
+isConst    := (name, symb) -> (isAlphaNumeric name
+    and not (isFunction or isType or isKeyword) (name, symb)
+    and (symb === symbol null or value symb =!= null))
+
+okay := method()
+okay(String, Keyword) :=
+okay(String, Symbol)  := (name, pkg) -> length name > 1 and isAlphaNumeric name
+
+cachedSymbols = new CacheTable
+STRINGS = format "///\\\\(/?/?[^/]\\\\|\\\\(//\\\\)*////[^/]\\\\)*\\\\(//\\\\)*///"
+
+-- This banner is added to the top of generated grammars
+banner := "Auto-generated for Macaulay2-@M2VERSION@. Do not modify this file manually."
+
+generateSymbols = () -> (
+    ----------------------------------------------------------------------------
+    -- Get a list of all symbols visible just after loading preloaded packages
+    allPkgNames := separate(" ", version#"packages") | {"Core"};
+    loadedPkgNames := join(Core#"preloaded packages",
+	{"Core", "Text", "Parsing", "SimpleDoc"});
+    symbols := unique (sortBy ascii @@ first) join(
+	apply(allPkgNames, pkgname -> (pkgname, symbol Core)),
+	flatten apply(loadedPkgNames, pkgname -> (
+		pkg := needsPackage pkgname;
+		select(pairs pkg.Dictionary, okay))));
+
+    if length symbols < 1500 then error "expected more entries for M2-symbols";
+
+    -------------------------------------------------------------------------------
+    -- Put the symbols into bins
+    cachedSymbols.Symbol   = first \ select(symbols,
+	(name, symb) ->isAlphaNumeric name);
+    cachedSymbols.Keyword  = first \ select(symbols, isKeyword);
+    cachedSymbols.Type     = first \ select(symbols, isType);
+    cachedSymbols.Function = first \ select(symbols, isFunction);
+    cachedSymbols.Constant = sort join(
+	first \ select(symbols, isConst), {
+	    -- SimpleDoc docstring keywords -- keep updated
+	    -- NodeFunctions
+	    "Node",
+	    "Synopsis",
+	    -- DescriptionFunctions
+	    "CannedExample",
+	    "Code",
+	    "Example",
+	    "Pre",
+	    "Tree",
+	    -- ConsequenceFunctions
+	    "Item"
+	    });
+    )
+
+-------------------------------------------------------------------------------
+-- Generate syntax files from templates in the same directory
+
+generateGrammar = method()
+generateGrammar(String, String, Function) := (template, outfile, demarkf) -> (
```

Comment:
I wonder if it would make more sense to replace `demarkf` with `tolistf`, which also takes care of how line wrapping long lists of words and how to enclose the list (e.g. braces, brackets, or nothing at all.